README.md
# Content Decay Detector

Compare two comparable windows of an exported metric and report the pages whose
recent window sits far enough below the baseline window to be worth a human
look. Optional link and freshness exports are reported alongside the
comparison, as context.

- **Repository:** [edilec/content-decay-detector](https://github.com/edilec/content-decay-detector)
- **Area:** Content & Publishing
- **License:** MIT

The hard part of a decay check is not the arithmetic. It is refusing to answer
when the data cannot support an answer, and refusing to dress an observation up
as a diagnosis. This tool does three things about that:

- **Below the configured minimum volume, a page is inconclusive.** Not "no
  decline found". The run is `incomplete` and exits 2, so a sparse page can
  never be mistaken for a healthy one.
- **The two windows must be comparable.** Same length, same number of days left
  after seasonality exclusions, and positioned the way the seasonality rule
  requires. A configuration that would compare unlike things is refused before
  any export is opened.
- **It never claims a cause.** It compares numbers somebody else exported. It
  has no view of a search engine, an edit, a redirect or a competitor, so it
  says nothing about ranking, position, penalties or why anything changed. That
  is enforced in code, not just in prose: a frozen `CAUSAL_TERMS` list and a
  tagged template that refuses a finding message containing one.

Nothing is fetched, ever. Every number it reasons about comes from a file
inside a declared input root.

## Install

Node 22 or newer. No runtime dependencies, no dev dependencies, Node built-ins
only.

```sh
npm install content-decay-detector
```

## Use

```sh
npx content-decay-detector --config examples/clean/decay.config.json
npx content-decay-detector --config site/decay.config.json --root site --json
npx content-decay-detector --config site/decay.config.json --minimum-volume 500
```

```
--config FILE        Comparison configuration (required)
--root DIR           Input root every declared export path resolves against and
                     may not escape. Defaults to the directory holding the
                     config.
--minimum-volume N   Override config.minimumVolume
--json               Suppress the human summary on stderr
-h, --help           Show help
```

stdout carries the JSON report and nothing else. stderr carries the human
summary and diagnostics.

| Exit | Meaning |
| ---: | --- |
| `0` | every page with enough data was compared and none declined past the threshold |
| `1` | the comparison completed and at least one page declined past it |
| `2` | invalid configuration, or evidence the comparison could not obtain |

Exit 2 has two shapes. A configuration error means the run never had a subject,
so stdout stays **empty** and the message goes to stderr. Evidence that could
not be read, decoded or scored means the run had a subject and failed to
establish something about it, so stdout carries a report with status
`incomplete` naming what was not established. A consumer piping stdout must
handle an empty stdout on exit 2.

## Configuration

```json
{
  "schemaVersion": "1",
  "metric": "sessions",
  "series": ["exports/analytics.json"],
  "freshness": ["exports/freshness.json"],
  "baseline": { "start": "2024-09-01", "end": "2024-09-30" },
  "recent":   { "start": "2025-09-01", "end": "2025-09-30" },
  "seasonality": {
    "rule": "year-over-year",
    "excludePeriods": [{ "start": "09-14", "end": "09-16" }]
  },
  "minimumVolume": 200,
  "minimumDays": 20,
  "maxCoverageGapDays": 2,
  "stalenessDays": 540,
  "thresholds": { "decayRatio": 0.7, "warnRatio": 0.9, "linkRatio": 0.7, "minimumLinks": 10 }
}
```

Every key is checked. An unknown key, including a one-character typo, is a
configuration error rather than a silently ignored setting.

- `seasonality.rule` is `none`, `adjacent` or `year-over-year`.
- `seasonality.excludePeriods` removes days from **both** windows. Spelled
  `YYYY-MM-DD` it removes those dates; spelled `MM-DD` it removes that calendar
  position in every year, which is how one yearly sale comes out of both
  windows at once.
- `minimumVolume` is the baseline-window total a page needs before it is scored
  at all.
- `minimumDays` and `maxCoverageGapDays` keep the two windows like for like.
- `thresholds.decayRatio` and `warnRatio` are the recent mean as a fraction of
  the baseline mean, at or under which a page is reported.

### Analytics export

```json
{
  "schemaVersion": "1",
  "metric": "sessions",
  "rows": [{ "page": "/guides/install", "date": "2025-09-01", "value": 41 }]
}
```

### Link and freshness export

```json
{
  "schemaVersion": "1",
  "pages": [
    {
      "page": "/guides/install",
      "lastModified": "2025-06-01",
      "inboundLinks": { "baseline": 45, "recent": 44 }
    }
  ]
}
```

Both are JSON. There is no CSV reader; convert your export first. That keeps one
parser, with one set of bounds, instead of two.

## Examples

```sh
node bin/content-decay-detector.mjs --config examples/clean/decay.config.json   # exit 0
node bin/content-decay-detector.mjs --config examples/broken/decay.config.json  # exit 2
```

The broken project shows both halves of the contract in one run: `/guides/install`
falls from about 62 to about 22 a day with 27 days of data in each window and is
flagged as an error, while `/blog/old-post` has 30 sessions across five baseline
days and is reported inconclusive rather than green. Because something was left
unestablished, the run as a whole is `incomplete`.

## Limits and non-goals

This tool reads numbers out of files and compares two windows of them. It
**cannot** conclude:

- **That anything caused the decline.** Not a search engine, not an edit, not a
  redirect, not a competitor, not a season it was not told about. It sees two
  columns of numbers. Any message implying otherwise is a bug, and the code
  refuses to build one.
- **Anything about search ranking or position.** It never sees a search result
  page. The words are not in its vocabulary.
- **That a decline matters.** A page that was always marginal and halved is
  arithmetically identical to a flagship that halved. Deciding which matters is
  your job.
- **That a page is the right unit.** It compares whatever page identifiers your
  export contains. If your export splits one article across three URLs, it
  compares three things.
- **That a flagged page is worse than an unflagged one.** There is no ordering
  of pages by badness here, and `declining` in the summary is a count, not a
  league table.
- **That an unflagged page is healthy.** A page can be below `minimumVolume`, or
  short of `minimumDays`, or absent from the export entirely, or present in the
  export with every row falling outside both comparison windows. The first two
  are reported as inconclusive. The last two are invisible: a page whose rows
  all sit outside the windows is not counted in `summary.checked`, produces no
  finding, and does not stop the run passing, however steeply its numbers moved
  outside them. Only rows inside the two windows are compared.
- **That a missing day means zero.** A date absent from the export is treated as
  no observation. If a page fell to literal zero and your export omits zero
  rows, this tool sees a coverage gap and says it cannot judge. Export zero rows
  explicitly if you want zeros counted.
- **That the two windows are truly comparable in the world.** It checks that
  they are the same length, keep the same days, and sit where the seasonality
  rule says. It cannot know that one of them contained an outage, a campaign, or
  a tracking change. `seasonality.excludePeriods` exists for the ones you know
  about; it cannot help with the ones you do not.
- **That your export is accurate.** Sampling, bot filtering, consent-gated
  analytics and tag changes are all invisible to it.
- **Anything at all when it could not read something.** An unreadable export, an
  undecodable one, an exceeded limit, a malformed row: each is reported, and
  each makes the run `incomplete` rather than passing on what was left. Numbers
  in an incomplete report describe the rows that were read, not the export.

It also never writes anything, never fetches anything, and refuses any
configured path that leaves the input root, by spelling or through a symbolic
link.

## Determinism

Findings sort by `(location.file, location.pointer, ruleId, message)`, each
compared by UTF-16 code unit; `localeCompare` is never used. No wall clock is
read — the only "now" the tool has is the end of the recent window, which the
configuration supplies. Two runs over identical inputs produce byte-identical
stdout.

## Verify

```sh
npm run check
```

`check` runs the linter, the test suite, the clean example, and a packaging dry
run. The suite covers the public API and the real CLI.

Every guarantee stated here and in [`docs/decay-rules.md`](./docs/decay-rules.md)
is defended by a test that fails when the guarantee is taken out of the source.
That is not a claim about intent: each guarantee was removed in turn and the
suite was run against the result.

The guarantees that decide a verdict are pinned by what a run emits, not by one
declaration agreeing with another. `test/severity.test.mjs` drives every rule in
the catalog through the real CLI and checks the emitted severity, the status and
the exit code against hand-written expectations, so a coordinated edit of the
severity table, the documentation and the test catalog cannot hide a downgrade.
`test/ordering.test.mjs` asserts the exact order of a real report over inputs
that code units and collation order differently, so substituting a collating
comparator fails it even though the source carries no `localeCompare`.
`test/boundaries.test.mjs` pins both sides of twenty documented limits, against
the 22 mutations that tighten each comparison by one.

## Documentation

- [`docs/decay-rules.md`](./docs/decay-rules.md) — the rule catalog, every
  severity, every limit, the comparability rules, and the boundaries.
- [`CHANGELOG.md`](./CHANGELOG.md) — release notes.

## License

MIT. See [LICENSE](./LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project are recorded here. The format follows
[Keep a Changelog](https://keepachangelog.com/en/1.1.0/) and this project uses
[semantic versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- Two-window comparison of an exported metric, with a configurable minimum
  baseline volume below which a page is reported inconclusive rather than
  scored. That distinction is the point of the tool: sparse data is never "no
  decline".
- Comparability checks decided from the configuration alone, before any export
  is opened: equal window length, equal included days after seasonality
  exclusions, a recent window that begins after the baseline window ends, and
  the positioning the configured seasonality rule demands.
- Seasonality exclusions spelled either as absolute dates or as an `MM-DD`
  calendar position that is removed from every year, so one yearly event comes
  out of both windows at once.
- Optional link and freshness exports, reported as context next to the metric
  comparison and never as an account of it.
- A frozen `ruleId -> severity` catalog of 25 rules, asserted against
  `docs/decay-rules.md` in both directions and against a third hand-written copy
  in the test suite.
- Six configurable limits and five fixed configuration bounds, each enforced and
  each reporting a named finding or a named configuration error rather than
  truncating silently.
- `content-decay-detector` CLI with `--config`, `--root`, `--minimum-volume`,
  `--json` and `--help`, emitting the v1 report envelope on stdout.
- A standing `disclaimer` on every report, whatever the verdict, saying that the
  tool establishes no cause, observes no search ranking or position, and does
  not conclude that anything is wrong.
- Clean and deliberately broken example projects under `examples/`.

### Fixed

- A parse failure no longer reproduces the file it failed on. `JSON.parse`
  reports a failure either by position or by quoting the input back —
  `Unexpected token 'A', "AKIAIOSFODNN7EXAMPLE" is not valid JSON`, which is the
  whole document when the document is short — and that quote sits at the front
  of the message, where sanitising and the length cut both leave it intact. An
  export or a configuration short enough to be only a credential was therefore
  published by the very message that failed to read it, through
  `series-unparsable`, `freshness-unparsable` and the config diagnostic on
  stderr. All three are built from the position, line and column alone now.
- The diagnostic for a configuration file that is not JSON quoted the parser's
  message unsanitised, and that message carries a slice of the document. A
  configuration whose bytes held `U+0085` or `U+202E` therefore reached stderr
  raw and could forge a line or reverse displayed text. It is sanitised like
  every other untrusted string now.

### Notes

- Causal and ranking language is refused in code, not only in review. Finding
  messages are built through a tagged template whose literals are checked
  against a frozen `CAUSAL_TERMS` list; the interpolated values, which come from
  untrusted export files, are sanitised instead, so a page genuinely named
  `/why-we-were-penalised` reports normally without the tool appearing to make
  the claim.
- Report status is derived from the findings rather than from a mutable flag,
  and the three rules whose only defence is their evidence-missing marking
  (`insufficient-volume`, `insufficient-coverage`, `window-coverage-mismatch`)
  each have an end-to-end test that fails if that marking is removed.
- Severity, the evidence markings and the sort order are pinned by what a run
  emits, not by one declaration agreeing with another: every rule in the catalog
  is driven through the CLI and its status and exit code checked against
  hand-written expectations, and the ordering tests assert the exact sequence of
  real reports over inputs that code-unit and collation ordering disagree about.
- `0.1.0` is the version recorded in `package.json`. No release has been
  published.


test/dates.test.mjs
import assert from 'node:assert/strict'
import test from 'node:test'

import {
  SEASONALITY_RULES,
  buildWindow,
  comparabilityProblems,
  dayToIso,
  parseIsoDate,
  parseMonthDay,
} from '../src/index.mjs'

const window = (start, end, periods = []) => {
  const from = parseIsoDate(start)
  const to = parseIsoDate(end)
  assert.equal(from.ok && to.ok, true, `${start}..${end} must be real dates`)
  return buildWindow(from.day, to.day, periods)
}

test('a calendar date is accepted only when it names a real day', () => {
  const ok = parseIsoDate('2025-09-01')
  assert.equal(ok.ok, true)
  assert.equal(dayToIso(ok.day), '2025-09-01')
  assert.equal(parseIsoDate('2024-02-29').ok, true, 'a leap day is a real day')

  for (const [value, fragment] of [
    ['2025-02-30', 'real calendar date'],
    ['2025-13-01', 'real calendar date'],
    ['2023-02-29', 'real calendar date'],
    ['2025-9-01', 'YYYY-MM-DD'],
    ['20250901', 'YYYY-MM-DD'],
    ['0025-01-01', '1000 or later'],
    [20250901, 'must be a string'],
    [null, 'must be a string'],
  ]) {
    const parsed = parseIsoDate(value)
    assert.equal(parsed.ok, false, `${String(value)} must be refused`)
    assert.match(parsed.reason, new RegExp(fragment, 'u'), String(value))
  }
})

test('day numbers are consecutive integers and round trip through ISO', () => {
  const first = parseIsoDate('2024-12-30')
  const second = parseIsoDate('2024-12-31')
  const third = parseIsoDate('2025-01-01')
  assert.equal(second.day - first.day, 1)
  assert.equal(third.day - second.day, 1)
  assert.equal(dayToIso(third.day), '2025-01-01')
})

test('a recurring MM-DD position is validated against real months', () => {
  assert.equal(parseMonthDay('11-25').key, 1125)
  assert.equal(parseMonthDay('02-29').ok, true, 'February 29 exists in some years')
  assert.equal(parseMonthDay('02-30').ok, false)
  assert.equal(parseMonthDay('13-01').ok, false)
  assert.equal(parseMonthDay('1-1').ok, false)
  assert.equal(parseMonthDay('2025-01-01').ok, false)
})

test('an absolute exclusion removes exactly the days it names', () => {
  const start = parseIsoDate('2025-09-10').day
  const end = parseIsoDate('2025-09-12').day
  const built = window('2025-09-01', '2025-09-30', [{ kind: 'absolute', startDay: start, endDay: end }])
  assert.equal(built.length, 30)
  assert.equal(built.excludedDays, 3)
  assert.equal(built.includedDays, 27)
  assert.equal(built.included.has(parseIsoDate('2025-09-11').day), false)
  assert.equal(built.included.has(parseIsoDate('2025-09-13').day), true)
})

test('a recurring exclusion removes the same calendar days in every year', () => {
  const period = { kind: 'recurring', startKey: 914, endKey: 916 }
  const baseline = window('2024-09-01', '2024-09-30', [period])
  const recent = window('2025-09-01', '2025-09-30', [period])
  assert.equal(baseline.includedDays, 27)
  assert.equal(recent.includedDays, 27)
  assert.equal(baseline.included.has(parseIsoDate('2024-09-15').day), false)
  assert.equal(recent.included.has(parseIsoDate('2025-09-15').day), false)
  assert.equal(recent.included.has(parseIsoDate('2025-09-17').day), true)
})

test('windows of different length are not comparable', () => {
  const problems = comparabilityProblems(window('2024-09-01', '2024-09-30'), window('2025-09-01', '2025-09-29'), 'none')
  assert.equal(problems.length > 0, true)
  assert.match(problems[0], /different length/u)
})

test('a leap day in only one window is refused rather than absorbed', () => {
  const baseline = window('2024-02-01', '2024-03-01')
  const recent = window('2025-02-01', '2025-03-01')
  assert.equal(baseline.length, 30)
  assert.equal(recent.length, 29)
  assert.match(comparabilityProblems(baseline, recent, 'year-over-year').join(' '), /different length/u)
})

test('an exclusion landing in only one window is not comparable', () => {
  const period = { kind: 'absolute', startDay: parseIsoDate('2025-09-10').day, endDay: parseIsoDate('2025-09-12').day }
  const baseline = window('2024-09-01', '2024-09-30', [period])
  const recent = window('2025-09-01', '2025-09-30', [period])
  assert.equal(baseline.includedDays, 30)
  assert.equal(recent.includedDays, 27)
  assert.match(comparabilityProblems(baseline, recent, 'year-over-year').join(' '), /only one window/u)
})

test('exclusions that remove every day leave nothing to compare', () => {
  const period = { kind: 'recurring', startKey: 901, endKey: 930 }
  const baseline = window('2024-09-01', '2024-09-30', [period])
  const recent = window('2025-09-01', '2025-09-30', [period])
  assert.equal(baseline.includedDays, 0)
  assert.match(comparabilityProblems(baseline, recent, 'none').join(' '), /removed every day/u)
})

test('the year-over-year rule demands the same month and day', () => {
  const aligned = comparabilityProblems(window('2024-09-01', '2024-09-30'), window('2025-09-01', '2025-09-30'), 'year-over-year')
  assert.deepEqual(aligned, [])
  const shifted = comparabilityProblems(window('2024-09-01', '2024-09-30'), window('2025-10-01', '2025-10-30'), 'year-over-year')
  assert.match(shifted.join(' '), /same month and day/u)
  assert.deepEqual(
    comparabilityProblems(window('2024-09-01', '2024-09-30'), window('2025-10-01', '2025-10-30'), 'none'),
    [],
    'the same pair is fine when no seasonality rule is configured',
  )
})

test('the adjacent rule demands the recent window start the next day', () => {
  const touching = comparabilityProblems(window('2025-07-01', '2025-07-30'), window('2025-07-31', '2025-08-29'), 'adjacent')
  assert.deepEqual(touching, [])
  const gapped = comparabilityProblems(window('2025-07-01', '2025-07-30'), window('2025-08-05', '2025-09-03'), 'adjacent')
  assert.match(gapped.join(' '), /the day after/u)
})

test('the recent window may not begin before the baseline window ends', () => {
  const overlapping = comparabilityProblems(window('2025-09-01', '2025-09-30'), window('2025-09-15', '2025-10-14'), 'none')
  assert.match(overlapping.join(' '), /must begin after/u)
})

test('the seasonality rules are exactly the three documented ones', () => {
  assert.deepEqual([...SEASONALITY_RULES], ['none', 'adjacent', 'year-over-year'])
})


src/rules.mjs
/**
 * The rule catalog, the severity table, and everything that turns findings into
 * a status.
 *
 * Three separate defences live here, and each one exists because its absence
 * has produced a green build over a real failure somewhere in this catalog:
 *
 * 1. Severity is declared exactly once, in `RULE_SEVERITY`. Every finding takes
 *    its severity from that table, and an unknown rule id throws rather than
 *    defaulting to something harmless.
 * 2. `status` is derived from the findings, not from a mutable flag. A run
 *    whose evidence was missing, sparse, truncated or undecodable is
 *    `incomplete`, and there is no single assignment whose deletion would let
 *    an unread or unscored input report `pass`.
 * 3. A finding's message must be built with the `msg` tagged template. The
 *    template's own literals are checked for causal and ranking language and
 *    the interpolated values -- which come from untrusted export files -- are
 *    sanitised. This tool observes a difference between two supplied numbers;
 *    it is not allowed to phrase that as a cause or as a ranking claim.
 */

/** Deterministic order: UTF-16 code unit, never locale collation. */
export function byCodeUnit(a, b) {
  return a === b ? 0 : a < b ? -1 : 1
}

export const SEVERITIES = Object.freeze(['error', 'warning', 'info'])

export const RULE_SEVERITY = Object.freeze({
  'decline-near-threshold': 'warning',
  'decline-over-threshold': 'error',
  'duplicate-observation': 'error',
  'freshness-duplicate-page': 'error',
  'freshness-not-utf8': 'error',
  'freshness-page-unknown': 'info',
  'freshness-record-invalid': 'error',
  'freshness-record-limit-exceeded': 'error',
  'freshness-too-large': 'error',
  'freshness-unparsable': 'error',
  'freshness-unreadable': 'error',
  'inbound-link-decline': 'warning',
  'insufficient-coverage': 'warning',
  'insufficient-volume': 'warning',
  'metric-mismatch': 'error',
  'no-pages-scored': 'error',
  'page-limit-exceeded': 'error',
  'row-invalid': 'error',
  'row-limit-exceeded': 'error',
  'series-not-utf8': 'error',
  'series-too-large': 'error',
  'series-unparsable': 'error',
  'series-unreadable': 'error',
  'stale-content': 'info',
  'window-coverage-mismatch': 'warning',
})

export const RULE_IDS = Object.freeze(Object.keys(RULE_SEVERITY).sort(byCodeUnit))

/**
 * Rules that mean the tool did not obtain the evidence a verdict would need.
 * Any one of them makes the whole report `incomplete` and the process exit 2,
 * whatever the rule's own severity happens to be.
 *
 * `insufficient-volume`, `insufficient-coverage` and `window-coverage-mismatch`
 * are the reason this list is not merely decoration. All three are `warning`
 * severity, so membership here is the only thing standing between a page the
 * tool could not judge and a green run. Sparse data is inconclusive; it is
 * never "no decline was found".
 */
export const EVIDENCE_MISSING_RULES = Object.freeze([
  'duplicate-observation',
  'freshness-duplicate-page',
  'freshness-not-utf8',
  'freshness-record-invalid',
  'freshness-record-limit-exceeded',
  'freshness-too-large',
  'freshness-unparsable',
  'freshness-unreadable',
  'insufficient-coverage',
  'insufficient-volume',
  'metric-mismatch',
  'no-pages-scored',
  'page-limit-exceeded',
  'row-invalid',
  'row-limit-exceeded',
  'series-not-utf8',
  'series-too-large',
  'series-unparsable',
  'series-unreadable',
  'window-coverage-mismatch',
].sort(byCodeUnit))

const EVIDENCE_MISSING_SET = new Set(EVIDENCE_MISSING_RULES)

export const EVIDENCE_LIMIT = 200
export const MAX_PAGE_ID_LENGTH = 256

export function severityFor(ruleId) {
  const severity = RULE_SEVERITY[ruleId]
  if (severity === undefined) throw new Error(`Unknown ruleId "${ruleId}"`)
  return severity
}

export function marksEvidenceMissing(ruleId) {
  severityFor(ruleId)
  return EVIDENCE_MISSING_SET.has(ruleId)
}

/**
 * Words this tool is not entitled to use about its own observations.
 *
 * It compares two windows of numbers that somebody else exported. It cannot see
 * a search engine, a competitor, an edit, or a redirect, so it may not say that
 * one thing produced another, and it may not talk about position in a result
 * set at all.
 */
export const CAUSAL_TERMS = Object.freeze([
  'because', 'cause', 'caused', 'causes', 'causing', 'due to', 'result of',
  'resulted in', 'results in', 'leads to', 'lead to', 'led to', 'triggered',
  'rank', 'ranks', 'ranked', 'ranking', 'rankings', 'position', 'positions',
  'penalty', 'penalties', 'penalise', 'penalised', 'penalize', 'penalized',
  'demoted', 'deindexed', 'punished', 'blame', 'fault', 'therefore', 'hence',
])

const CAUSAL_PATTERN = new RegExp(
  `\\b(?:${CAUSAL_TERMS.map((term) => term.replace(/[.*+?^${}()|[\]\\]/gu, '\\$&')).join('|')})\\b`,
  'iu',
)

export function findCausalClaim(text) {
  const match = CAUSAL_PATTERN.exec(String(text))
  return match === null ? null : match[0]
}

export function assertNoCausalClaim(text, what) {
  const term = findCausalClaim(text)
  if (term !== null) {
    throw new Error(`${what} may not claim cause or search position: "${term}" in ${JSON.stringify(String(text))}`)
  }
}

/**
 * U+2028 and U+2029 are written here as code points rather than as escape text
 * so that no editor, transfer or tool can quietly turn the escape into the
 * character it names.
 */
export const LINE_SEPARATORS = String.fromCharCode(0x2028, 0x2029)

const UNSAFE_CHARACTERS = new RegExp('[\\p{Cc}\\p{Cf}' + LINE_SEPARATORS + ']', 'gu')

/**
 * A bounded, control-character-free rendering of an untrusted string.
 *
 * Page identifiers, file names and metric names all arrive from export files
 * and all reach the report and the human summary. An identifier carrying a
 * newline forged extra lines in a shipped tool's human report, so every
 * untrusted string is flattened here -- not only the `evidence` field.
 */
export function sanitize(value, limit = EVIDENCE_LIMIT) {
  const flat = String(value)
    .replace(UNSAFE_CHARACTERS, ' ')
    .replace(/\s+/gu, ' ')
    .trim()
  return flat.length > limit ? `${flat.slice(0, limit - 3)}...` : flat
}

/**
 * What a `JSON.parse` failure may say about a file this tool did not write.
 *
 * V8 reports a parse failure two ways, and one of them quotes the input back:
 * `Unexpected token 'A', "AKIAIOSFODNN7EXAMPLE" is not valid JSON`. An export
 * or a config short enough to be only a credential is therefore reproduced in
 * full by its own error message, and `sanitize` does not stop it: that strips
 * control characters and cuts from the end, while the quoted input sits at the
 * front.
 *
 * The position is the useful half and carries no content, so it is kept
 * whenever V8 offers one. The quoted half never leaves this function.
 */
export function parseFailureDetail(error) {
  const message = String(error?.message ?? 'could not be parsed')
  const position = /at position \d+(?: \(line \d+ column \d+\))?/u.exec(message)
  if (position) return message.slice(0, position.index + position[0].length)
  const token = /^Unexpected token (.+?), ".*?"(?:\.\.\.)? is not valid JSON$/su.exec(message)
  if (token) return `unexpected token ${token[1]} at the start of the document`
  if (/^Unexpected end of JSON input$/u.test(message)) return message
  return 'the document could not be parsed as JSON'
}

/** A message whose literals have been checked and whose values are sanitised. */
export class SafeMessage {
  constructor(text) {
    this.text = text
    Object.freeze(this)
  }

  toString() {
    return this.text
  }
}

/**
 * Build a finding message.
 *
 * The tagged-template split is the point: `strings` is this tool's own voice
 * and is checked for causal and ranking language, while `values` come from
 * input files and are only sanitised. A page named `/why-we-were-penalised`
 * must not crash the run, and a sentence this tool wrote saying a decline was
 * caused by something must not ship.
 */
export function msg(strings, ...values) {
  let out = ''
  for (let index = 0; index < strings.length; index += 1) {
    assertNoCausalClaim(strings[index], 'A finding message')
    out += strings[index]
    if (index < values.length) out += sanitize(values[index])
  }
  return new SafeMessage(out)
}

export function at(file, pointer) {
  const location = {}
  if (file !== null && file !== undefined) location.file = file
  if (pointer !== null && pointer !== undefined) location.pointer = pointer
  return location
}

/** A documented field path for a page, with JSON Pointer escaping applied. */
export function pointerForPage(page) {
  const escaped = sanitize(page, MAX_PAGE_ID_LENGTH).replace(/~/gu, '~0').replace(/\//gu, '~1')
  return `/pages/${escaped}`
}

export function makeFinding(ruleId, message, location, extra = {}) {
  if (!(message instanceof SafeMessage)) {
    throw new Error(`Finding "${ruleId}" must build its message with the msg tagged template`)
  }
  const finding = { ruleId, severity: severityFor(ruleId), message: message.text, location }
  if (extra.evidence !== undefined) finding.evidence = sanitize(extra.evidence)
  if (extra.suggestion !== undefined) {
    assertNoCausalClaim(extra.suggestion, 'A finding suggestion')
    finding.suggestion = extra.suggestion
  }
  return finding
}

/** Findings sort by (file, pointer, ruleId, message), each by code unit. */
export function compareFindings(a, b) {
  return (
    byCodeUnit(a.location.file ?? '', b.location.file ?? '')
    || byCodeUnit(a.location.pointer ?? '', b.location.pointer ?? '')
    || byCodeUnit(a.ruleId, b.ruleId)
    || byCodeUnit(a.message, b.message)
  )
}

export function sortFindings(findings) {
  return [...findings].sort(compareFindings)
}

/**
 * Status is a function of the findings alone.
 *
 * Missing evidence outranks everything, including an error: a run that read
 * half its inputs has not established that the half it did read is the whole
 * story. There is no flag to delete.
 */
export function statusFor(findings) {
  for (const finding of findings) {
    if (EVIDENCE_MISSING_SET.has(finding.ruleId)) return 'incomplete'
  }
  for (const finding of findings) {
    if (finding.severity === 'error') return 'fail'
  }
  return 'pass'
}
